### DISCUSSION TITLE
Shape checking

### DISCUSSION THREAD

**Discussion by AriMKatz** [modular/modular#359] [Q&A] [answered]
Any plans for compile time array shape checking? There are several projects working with restricted dependent types to allow for something like this in a tractable manner. Does the mojo team intend on something similar?

**Comment by lattner** [comment, marked as answer]
I don't think it will be a feature of the default array type (because it needs to be dynamically resizable), but you can already express shape algebra in types using the parameter system, e.g. you can express types like this:

```
struct Parameterized[A: Int]:
  fn method[B: Int](self, other: Parameterized[B]) -> Parameterized[A+B]: ...
```

**Reply by AriMKatz** [reply]
That's cool
